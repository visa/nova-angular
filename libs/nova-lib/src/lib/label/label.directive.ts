/**
 *              Copyright (c) 2025 Visa, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *         http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 **/
import { BooleanInput, coerceBooleanProperty } from '@angular/cdk/coercion';
import { Directive, Input, HostBinding, ElementRef, HostListener } from '@angular/core';
import { UUIDService } from '../_utilities/services/uuid.service';

@Directive({
  standalone: true,
  selector: '[v-label]'
})
export class LabelDirective {
  /**
   * Attribute set to the id of the element it's labeling. <br />
   * This property is set by default for radio and checkbox components when used within <code>v-input-container</code> as directed.
   */
  @Input()
  get for(): string {
    return this._for;
  }
  set for(value: string) {
    this._for = value;
  }
  private _for: string;

  @HostBinding('attr.for')
  get labelFor(): string {
    return this._for;
  }

  /**
   * Sets custom id.
   * @default uuidService.getUUID('v-label-')
   * @builtin true
   */
  @Input() id: string = this.uuidService.getUUID('v-label-');
  @HostBinding('attr.id')
  get hostId(): string {
    return this.id;
  }

  /**
   * Provides custom class&#40;es&#41; for custom styling.
   * @default .v-label - when <code>progressLabel</code> is false and not used within a Circular Progress component.
   * @default .v-progress-label when <code>progressLabel</code> is true or when used within a Circular Progress component.
   */
  @Input() class: string = '';
  @HostBinding('class')
  get hostClasses(): string {
    this.class = [this.progressLabel ? 'v-progress-label' : 'v-label'].join(' ');
    return this.class;
  }

  /**
   * Swaps <code>v-label</code> class for <code>v-progress-label</code> when true. <br />
   * Intended for use when label is describing a progress component. <br />
   * Automatically set to true when used within a Circular Progress component.
   * @default false
   */
  @Input()
  get progressLabel(): boolean {
    return this._progressLabel;
  }
  set progressLabel(value: BooleanInput) {
    this._progressLabel = coerceBooleanProperty(value);
  }
  private _progressLabel: boolean = false;

  constructor(
    public el: ElementRef,
    private uuidService: UUIDService
  ) {} // needed for vdsChip styling

  @HostListener('click', ['$event'])
  onClick(event: Event) {
    // stops click event for firing for label **and** input.
    // event will still fire for input
    event.stopPropagation();
  }
}
