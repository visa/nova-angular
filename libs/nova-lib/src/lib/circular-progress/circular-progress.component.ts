/**
 *              Copyright (c) 2025 Visa, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *         http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 **/
import { BooleanInput, coerceBooleanProperty, coerceNumberProperty, NumberInput } from '@angular/cdk/coercion';
import { CommonModule } from '@angular/common';
import {
  AfterContentInit,
  AfterViewInit,
  Component,
  computed,
  ContentChild,
  ElementRef,
  HostBinding,
  Input,
  signal,
  ViewChild
} from '@angular/core';
import { AppReadyService } from '../_utilities/services/app-stable-check.service';
import { LabelDirective } from '../label/label.directive';

@Component({
  standalone: true,
  imports: [CommonModule],
  selector: '[v-progress-circular], [v-circular-progress]',
  templateUrl: './circular-progress.component.html'
})
export class CircularProgressComponent implements AfterContentInit, AfterViewInit {
  _percentageSignal = signal(this.percentage);
  _initialSize = signal(0); // circumference
  _dashOffset = computed(() => {
    if (this._initialSize() > 0) {
      return this._initialSize() - this._initialSize() * (this._percentageSignal() * 0.01);
    } else return 0;
  });

  @ViewChild('progressBar', { static: false }) progressBar: ElementRef;
  @ContentChild(LabelDirective) label: LabelDirective;

  /**
   * Provides custom class&#40;es&#41; for custom styling.
   * @default .v-progress.v-progress-circular
   */
  @Input() class: string = '';
  @HostBinding('class')
  get hostClasses(): string {
    return [
      this.class,
      'v-progress',
      'v-progress-circular',
      this.determinate ? '' : 'v-progress-indeterminate',
      this.small ? 'v-progress-circular-small' : '',
      this.determinate && this.percentage === 100 ? 'v-progress-complete' : '',
      this.invalid ? 'v-progress-error' : ''
    ].join(' ');
  }

  /**
   * Sets progress to determinate when true.
   * @default false
   */
  @Input()
  get determinate(): boolean {
    return this._determinate;
  }
  set determinate(value: BooleanInput) {
    this._determinate = coerceBooleanProperty(value);
  }
  _determinate: boolean = false;

  /**
   * Sets progress to invalid when true.
   * @default false
   */
  @Input()
  get invalid(): boolean {
    return this._invalid;
  }
  set invalid(value: BooleanInput) {
    this._invalid = coerceBooleanProperty(value);
  }
  _invalid: boolean = false;

  /**
   * Sets custom role.
   * @default 'progressbar' when determinate
   * @default null when indeterminate
   * @builtin true
   */
  @Input() role: string | void | null = null;
  @HostBinding('attr.role')
  get hostRole(): string | void | null {
    // If role is set, use it
    if (this.role) {
      return this.role;
    }
    if (this._determinate) {
      return 'progressbar';
    }
    return null;
  }

  /**
   * Sets the percentage for a <code>determinate</code> circular progress.
   * @default 0
   */
  @Input()
  get percentage(): number {
    return this._percentage;
  }
  set percentage(value: NumberInput) {
    this._percentage = coerceNumberProperty(value);
    this._percentageSignal.set(coerceNumberProperty(value));
  }
  _percentage: number = 0;

  /**
   * Sets progress to small variant when true.
   * @default false
   */
  @Input()
  get small(): boolean {
    return this._small;
  }
  set small(value: BooleanInput) {
    this._small = coerceBooleanProperty(value);
  }
  _small: boolean = false;

  /**
   * Aria attribute pointing to id of labelling element.
   * @default &lt;child-label-id&gt;
   * @builtin true
   */
  @Input('aria-labelledby') ariaLabelledby: string;
  @HostBinding('attr.aria-labelledby')
  get hostAriaLabelledby(): string {
    return this.ariaLabelledby;
  }

  constructor(private appReadyService: AppReadyService) {}

  ngAfterContentInit(): void {
    if (this.label) {
      this.label.progressLabel = true;

      if (!this.ariaLabelledby && this.label.id) {
        this.ariaLabelledby = this.label.id;
      }
    }
  }

  ngAfterViewInit(): void {
    if (this.progressBar) {
      let radius: number; // default in case isPlatformBrowser fails
      // best guess of radius if radius is not available
      // by default, small size is 24
      // by default, regular size is 48
      const defaultSize = this.small ? 48 : 72;
      setTimeout(() => {
        // .getPropertyValue('r') not working outside of setTimeout
        if (this.appReadyService.checkIsPlatformBrowser()) {
          radius = parseInt(window.getComputedStyle(this.progressBar.nativeElement).getPropertyValue('r'));
        }
        // by default, radius is half of the default size minus 4 - the stroke width
        radius = isNaN(radius) ? (defaultSize - 4) / 2 : radius;
        this._initialSize.set(2 * Math.PI * radius);
      }, 0);
    }
  }
}
