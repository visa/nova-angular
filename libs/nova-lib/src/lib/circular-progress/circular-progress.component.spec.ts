/**
 *              Â© 2025 Visa
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *         http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 **/
import { Component, ViewChild } from '@angular/core';
import { ComponentFixture, TestBed } from '@angular/core/testing';
import { render } from '@testing-library/angular';
import { LabelDirective } from '../label/label.directive';
import { CircularProgressComponent } from './circular-progress.component';

@Component({
  template: `<div v-progress-circular>
    <label v-label>Label</label>
  </div> `
})
class TestCircularProgressComponent {
  @ViewChild(CircularProgressComponent) progress: CircularProgressComponent;
}

describe('CircularProgressComponent', () => {
  describe('class', () => {
    let component: TestCircularProgressComponent;
    let fixture: ComponentFixture<TestCircularProgressComponent>;
    let htmlEl: HTMLElement;

    beforeEach(async () => {
      await TestBed.configureTestingModule({
        declarations: [TestCircularProgressComponent],
        imports: [CircularProgressComponent, LabelDirective]
      }).compileComponents();
    });

    beforeEach(() => {
      fixture = TestBed.createComponent(TestCircularProgressComponent);
      component = fixture.componentInstance;
      htmlEl = fixture.nativeElement.querySelector('[v-progress-circular]');
      fixture.detectChanges();
    });

    it('should exist', () => {
      expect(component).toBeTruthy();
    });

    describe('Input properties', () => {
      it(`should have 'v-progress' class by default`, () => {
        expect(htmlEl.classList.toString()).toEqual(expect.stringContaining('v-progress'));
      });

      it(`should have 'v-progress-circular' class by default`, () => {
        expect(htmlEl.classList.toString()).toEqual(expect.stringContaining('v-progress-circular'));
      });

      it('should add custom class', () => {
        component.progress.class = 'v-progress-custom';
        fixture.detectChanges();
        expect(htmlEl.classList.toString()).toEqual(expect.stringContaining('v-progress-custom'));
      });

      it('should be indeterminate by default', () => {
        expect(component.progress.determinate).toBeFalsy();
      });

      it('should have indeterminate class by default', () => {
        expect(htmlEl.classList.toString()).toEqual(expect.stringContaining('v-progress-indeterminate'));
      });

      it('should not have indeterminate class when determinate is true', () => {
        component.progress.determinate = true;
        fixture.detectChanges();
        expect(htmlEl.classList.toString()).toEqual(expect.not.stringContaining('v-progress-indeterminate'));
      });

      it('should have the complete class when determinate is true and percentage is max', () => {
        component.progress.percentage = 100;
        component.progress.determinate = true;
        fixture.detectChanges();
        expect(htmlEl.classList.toString()).toEqual(expect.stringContaining('v-progress-complete'));
      });

      it('should have small class when small', () => {
        component.progress.small = true;
        fixture.detectChanges();
        expect(htmlEl.classList.toString()).toEqual(expect.stringContaining('v-progress-circular-small'));
      });

      it('should have error class when invalid', () => {
        component.progress.invalid = true;
        fixture.detectChanges();
        expect(htmlEl.classList.toString()).toEqual(expect.stringContaining('v-progress-error'));
      });

      it('should allow custom role', () => {
        component.progress.role = 'tab';
        fixture.detectChanges();
        expect(htmlEl.getAttribute('role')).toBe('tab');
      });
    });

    describe('the template', () => {
      it(`should have 'v-progress-circular-track' on inner svg`, () => {
        const childSvg = htmlEl.querySelector('svg');
        if (childSvg)
          expect(childSvg.classList.toString()).toEqual(expect.stringContaining('v-progress-circular-track'));
      });

      it(`should have 'v-progress-circular-background' on inner first circle child`, () => {
        const firstCircle = htmlEl.querySelectorAll('circle')[0];
        if (firstCircle)
          expect(firstCircle.classList.toString()).toEqual(expect.stringContaining('v-progress-circular-background'));
      });

      it(`should have 'v-progress-circular-bar' on inner second circle child`, () => {
        const firstCircle = htmlEl.querySelectorAll('circle')[1];
        if (firstCircle)
          expect(firstCircle.classList.toString()).toEqual(expect.stringContaining('v-progress-circular-bar'));
      });
    });

    describe('the progress label', () => {
      it('should identify a child LabelDirective', () => {
        expect(component.progress.label).toBeTruthy();
      });

      it('should set the aria-labelledby to the labels ID by default', () => {
        expect(htmlEl.getAttribute('aria-labelledby')).toEqual(component.progress.label.id);
      });

      it('should allow for a custom aria-labelledby attribute if given', () => {
        component.progress.ariaLabelledby = 'custom description';
        fixture.detectChanges();
        expect(htmlEl.getAttribute('aria-labelledby')).toEqual('custom description');
      });
    });

    xdescribe('computed styles', () => {
      // I'd like to test computed styles based on what we get for radius, etc, but I'm unable to get it working
    });
  });

  describe('rendering', () => {
    it('should allow for custom sizes', async () => {
      jest.useFakeTimers();
      const { container } = await render(
        `<style>
          .custom-circular-progress {
            transform: scale(0.25);
            block-size: calc(var(--v-progress-circular-size) * 0.25);
            inline-size: calc(var(--v-progress-circular-size) * 0.25);
            display: grid;
            justify-content: center;
            align-content: center;

          }
        </style>
        <div v-progress-circular class="custom-circular-progress">
          <span role="alert" vSR>Loading...</span>
        </div>
        `,
        {
          imports: [CircularProgressComponent, LabelDirective]
        }
      );
      jest.runAllTimers();
      expect(container).toMatchSnapshot();
      jest.useRealTimers();
    });
  });
});
