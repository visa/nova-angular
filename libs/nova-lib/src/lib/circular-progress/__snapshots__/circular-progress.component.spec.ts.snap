// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CircularProgressComponent rendering should allow for custom sizes 1`] = `
<div
  id="root17"
>
  <div
    class="custom-circular-progress v-progress v-progress-circular v-progress-indeterminate"
    v-progress-circular=""
  >
    <svg
      class="v-progress-circular-track"
    >
      <circle
        class="v-progress-circular-background"
      />
      <circle
        class="v-progress-circular-bar"
      />
    </svg>
    <span
      role="alert"
      vsr=""
    >
      Loading...
    </span>
  </div>
</div>
`;
